### DISCUSSION TITLE
How to decode the output tensor?

### DISCUSSION THREAD

**Discussion by gronnmann** [ultralytics/yolov5#9948] [Q&A]
Hey!
I have been doing inference on a YOLOv5-exported model on TensorRT using C++.
The model loads correctly, and the inference works, but now I'm left with a 1x25200x85 tensor.
I'm wondering - how do I make use of the data inside of that tensor, how is it located?

**Comment by spacewalk01** [comment]
25200 is the number of detected candidate boxes. So, in order to get the final output you need to apply nms (non-max suppression).
85 consists of 80 class scores and 4 box locations (x,y,width,height) and 1 objectness score. Objectness score will help you to identify if this box is an actual object and from 80 classes, you can figure out which class it belongs to, and (x,y,width, height) location tells you where your object locates. You need also to copy your tensor data in CUDA to CPU memory for further postprocessing. Here is an example I used: [link is here](https://github.com/spacewalk01/tensorrt-yolov6/blob/master/trt_demo/SampleYolo.cpp)
```
std::vector<BoundingBox> SampleYolo::get_bboxes(int batch_size, int keep_topk, float *output)
{
    std::vector<BoundingBox> bboxes;

    std::vector<int> classIds;
    std::vector<float> confidences;
    std::vector<cv::Rect> rectBoxes;

    int i = 0;
    int nc = 80;
    while (i < 8400)
    {
        // Box
        int k = i * 85;
        float object_conf = output[k + 4];

        if (object_conf < this->mParams.confThreshold)
        {
            i++;
            continue;
        }

        // (center x, center y, width, height) to (x, y, w, h)
        float x = (output[k] - output[k + 2] / 2);
        float y = (output[k + 1] - output[k + 3] / 2);
        float width = output[k + 2];
        float height = output[k + 3];

        // Classes
        float class_conf = output[k + 5];
        int classId = 0;

        for (int j = 1; j < nc; j++)
        {
            if (class_conf < output[k + 5 + j])
            {
                classId = j;
                class_conf = output[k + 5 + j];
            }
        }
        
        i++;
        
        class_conf *= object_conf;

        classIds.push_back(classId);
        confidences.push_back(class_conf);
        rectBoxes.emplace_back(cv::Rect((int)x, (int)y, (int)width, (int)height));

    }

    // Non-maximum suppression to eliminate redudant overlapping boxes
    std::vector<int> indices;
    cv::dnn::NMSBoxes(rectBoxes, confidences, this->mParams.confThreshold, this->mParams.nmsThreshold, indices);
    bboxes.reserve(indices.size());

    for (size_t i = 0; i < indices.size(); ++i)
    {
        BoundingBox box;
        // (x, y, w, h) to (x1, y1, x2, y2)
        box.x = rectBoxes[indices[i]].x;
        box.y = rectBoxes[indices[i]].y;
        box.w = rectBoxes[indices[i]].width;
        box.h = rectBoxes[indices[i]].height;
        box.score = confidences[indices[i]];
        box.cls = classIds[indices[i]];

        bboxes.emplace_back(box);
    }

    return bboxes;
}
```

**Reply by gronnmann** [reply]
Thanks a lot, exactly what I had spent hours looking for! I'll try it and let you know how that goes :)

**Reply by jappoz** [reply]
> 25200 is the number of detected candidate boxes. So, in order to get the final output you need to apply nms (non-max suppression). 85 consists of 80 class scores and 4 box locations (x,y,width,height) and 1 objectness score. Objectness score will help you to identify if this box is an actual object and from 80 classes, you can figure out which class it belongs to, and (x,y,width, height) location tells you where your object locates. You need also to copy your tensor data in CUDA to CPU memory for further postprocessing. Here is an example I used: [link is here](https://github.com/spacewalk01/tensorrt-yolov6/blob/master/trt_demo/SampleYolo.cpp)
> 
> ```
> std::vector<BoundingBox> SampleYolo::get_bboxes(int batch_size, int keep_topk, float *output)
> {
>     std::vector<BoundingBox> bboxes;
> 
>     std::vector<int> classIds;
>     std::vector<float> confidences;
>     std::vector<cv::Rect> rectBoxes;
> 
>     int i = 0;
>     int nc = 80;
>     while (i < 8400)
>     {
>         // Box
>         int k = i * 85;
>         float object_conf = output[k + 4];
> 
>         if (object_conf < this->mParams.confThreshold)
>         {
>             i++;
>             continue;
>         }
> 
>         // (center x, center y, width, height) to (x, y, w, h)
>         float x = (output[k] - output[k + 2] / 2);
>         float y = (output[k + 1] - output[k + 3] / 2);
>         float width = output[k + 2];
>         float height = output[k + 3];
> 
>         // Classes
>         float class_conf = output[k + 5];
>         int classId = 0;
> 
>         for (int j = 1; j < nc; j++)
>         {
>             if (class_conf < output[k + 5 + j])
>             {
>                 classId = j;
>                 class_conf = output[k + 5 + j];
>             }
>         }
>         
>         i++;
>         
>         class_conf *= object_conf;
> 
>         classIds.push_back(classId);
>         confidences.push_back(class_conf);
>         rectBoxes.emplace_back(cv::Rect((int)x, (int)y, (int)width, (int)height));
> 
>     }
> 
>     // Non-maximum suppression to eliminate redudant overlapping boxes
>     std::vector<int> indices;
>     cv::dnn::NMSBoxes(rectBoxes, confidences, this->mParams.confThreshold, this->mParams.nmsThreshold, indices);
>     bboxes.reserve(indices.size());
> 
>     for (size_t i = 0; i < indices.size(); ++i)
>     {
>         BoundingBox box;
>         // (x, y, w, h) to (x1, y1, x2, y2)
>         box.x = rectBoxes[indices[i]].x;
>         box.y = rectBoxes[indices[i]].y;
>         box.w = rectBoxes[indices[i]].width;
>         box.h = rectBoxes[indices[i]].height;
>         box.score = confidences[indices[i]];
>         box.cls = classIds[indices[i]];
> 
>         bboxes.emplace_back(box);
>     }
> 
>     return bboxes;
> }
> ```

What does the number 8400 in 

`while (i < 8400)`

represent?
